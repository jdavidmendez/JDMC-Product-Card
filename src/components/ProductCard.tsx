import styles from "./../styles/styles.module.css";
import { createContext } from "react";
import {
  ProductCardProps,
  ProductContextProps,
} from "../interfaces/interfaces";
import { useProduct } from "../hooks/useProduct";
import React from "react";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues,
}: ProductCardProps) => {
  const { counter, increaseBy, maxCount, isMaxReached, reset } = useProduct({ onChange, product, value, initialValues });
  return (
    <Provider
      value={{
        counter,
        increaseBy,
        maxCount,
        product,
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxReached,
          maxCount: initialValues?.maxCount,
          product,
          increaseBy,
          reset
        })}
      </div>
    </Provider>
  );
};
