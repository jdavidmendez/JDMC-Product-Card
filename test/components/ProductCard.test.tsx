import renderer, { act } from 'react-test-renderer';
import React from 'react';
import { product1 } from '../data/products';
import ProductCard from '../../src/components/index';


describe('pruebas en el productCard', () => {
    test('debe mostrar el componente correctamente', () => {
        const wrapper = renderer.create(
            <ProductCard product={product1} className='Custom class'>
                {
                    () => (<p>children</p>)
                }
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    test('Debe incrementar el contador', () => {
        const wrapper = renderer.create(
            <ProductCard product={product1} className='Custom class'>
                {
                    ({count,increaseBy}) => (
                        <>
                            <h1>Product card</h1>
                            <span>{count}</span>
                            <button onClick={()=> increaseBy(1)}></button>
                        </>
                    )
                }
            </ProductCard>
        )
        let tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
    
        act(() => {
            (tree as any).children[2].props.onClick();
        })

        tree = wrapper.toJSON();
        expect((tree as any).children[1].children[0]).toBe("1");
    })
})