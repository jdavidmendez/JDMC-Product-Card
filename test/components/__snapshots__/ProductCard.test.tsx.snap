// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el productCard Debe incrementar el contador 1`] = `
<div
  className="productCard Custom class"
>
  <h1>
    Product card
  </h1>
  <span>
    0
  </span>
  <button
    onClick={[Function]}
  />
</div>
`;

exports[`pruebas en el productCard debe mostrar el componente correctamente 1`] = `
<div
  className="productCard Custom class"
>
  <p>
    children
  </p>
</div>
`;
